src/rust/engine/options/src/parse.rs

```diff
@@ -274,49 +278,188 @@ fn format_parse_error(
+        option_value_parser::float_list_edits(value)
+            .map_err(|e| format_parse_error("float list", value, e))
+    }
+}
+
+impl Parseable for String {
+    fn parse(value: &str) -> Result<String, ParseError> {
+        Ok(value.to_owned())
+    }
+
+    fn parse_list(value: &str) -> Result<Vec<ListEdit<String>>, ParseError> {
+        option_value_parser::string_list_edits(value)
+            .map_err(|e| format_parse_error("string list", value, e))
+    }
+}
+
+pub(crate) trait ListMember: Parseable + DeserializeOwned {}
+
+impl ListMember for bool {}
+
+impl ListMember for i64 {}
+
+impl ListMember for f64 {}
+
+impl ListMember for String {}
+
+// If the corresponding unexpanded value points to a @fromfile, then the
+// first component is the path to that file, and the second is the value from the file,
+// or None if the file doesn't exist and the @?fromfile syntax was used.
+//
+// Otherwise, the first component is None and the second is the original value.
+type ExpandedValue = (Option<PathBuf>, Option<String>);
+
+fn mk_parse_err(err: impl Display, path: &Path) -> ParseError {
+    ParseError::new(format!(
+        "Problem reading {path} for {{name}}: {err_msg}",
+        path = path.display(),
+        err_msg = err,
+    ))
 }
 
-#[allow(dead_code)]
-pub(crate) fn parse_bool_list(value: &str) -> Result<Vec<ListEdit<bool>>, ParseError> {
-    option_value_parser::bool_list_edits(value)
-        .map_err(|e| format_parse_error("bool list", value, e))
+fn maybe_expand(value: String) -> Result<ExpandedValue, ParseError> {
+    if let Some(suffix) = value.strip_prefix('@') {
+        if suffix.starts_with('@') {
+            // @@ escapes the initial @.
+            Ok((None, Some(suffix.to_owned())))
+        } else {
+            match suffix.strip_prefix('?') {
+                Some(subsuffix) => {
+                    // @? means the path is allowed to not exist.
+                    let path = PathBuf::from(subsuffix);
+                    match fs::read_to_string(&path) {
+                        Ok(content) => Ok((Some(path), Some(content))),
+                        Err(err) if err.kind() == io::ErrorKind::NotFound => Ok((Some(path), None)),
+                        Err(err) => Err(mk_parse_err(err, &path)),
+                    }
+                }
+                _ => {
+                    let path = PathBuf::from(suffix);
+                    let content = fs::read_to_string(&path).map_err(|e| mk_parse_err(e, &path))?;
+                    Ok((Some(path), Some(content)))
+                }
+            }
+        }
+    } else {
+        Ok((None, Some(value)))
+    }
 }
 
-#[allow(dead_code)]
-pub(crate) fn parse_int_list(value: &str) -> Result<Vec<ListEdit<i64>>, ParseError> {
-    option_value_parser::int_list_edits(value).map_err(|e| format_parse_error("int list", value, e))
+pub(crate) fn expand(value: String) -> Result<Option<String>, ParseError> {
```

Comment:
This does no type conversion, just expands "path to @fromfile" -> "content of @fromfile". See usage in args/env/config.rs.